File: src/main/java/me/saro/kit/CacheStore.java
package me.saro.kit;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;


/**
 * @author PARK Yong Seo
 * @param <T>
 * @param <ID>
 * @Since 1.0.0
 */
public class CacheStore<ID, T> {

    private Map<ID, CacheWrapper<T>> store = new ConcurrentHashMap<>();

    final private long cacheTimeMillis;

    public CacheStore(long cacheTimeMillis) {
        this.cacheTimeMillis = cacheTimeMillis;
    }

    /**
     * get
     * @param id id
     * @param orElse or else value
     * @return
     */
    public synchronized T get(ID id, Function<ID, T> orElse) {
        return Optional
                .ofNullable(store.get(id))
                .map(CacheWrapper::get)
                .orElseGet(() -> {
                    final var data = orElse.apply(id);
                    store.put(id, new CacheWrapper<>(cacheTimeMillis, data));
                    return data;
                });
    }

    /**
     * get after forced update
     * @param id id
     * @param value update value
     * @return
     */
    public T getAfterForcedUpdate(ID id, T value) {
        store.put(id, new CacheWrapper<>(cacheTimeMillis, value));
        return value;
    }

    /**
     * remove
     * @param id id
     */
    public void remove(ID id) {
        store.remove(id);
    }

    /**
     * clear
     */
    public void clear() {
        store.clear();
    }

    public static class CacheWrapper<T> {

        final long expireCacheTimeMillis;
        final T data;

        public CacheWrapper(long cacheTimeMillis, T data) {
            this.expireCacheTimeMillis = System.currentTimeMillis() + cacheTimeMillis;
            this.data = data;
        }

        public T get() {
            return expireCacheTimeMillis >= System.currentTimeMillis() ? data : null;
        }
    }
}

File: src/main/kotlin/me/saro/kit/CacheStore.kt
package me.saro.kit

import java.util.concurrent.ConcurrentHashMap
import java.util.function.BiConsumer

/**
 * cache store
 * @author PARK Yong Seo
 */
class CacheStore<ID, T>(
    private val cacheTimeMillis: Long
) {
    private val store: MutableMap<ID, CacheWrapper<T>> = ConcurrentHashMap()

    /**
     * find by id
     * if does not exist update and return
     */
    fun find(id: ID, orElseUpdateAndGet: (ID) -> T): T = store[id]?.data ?: update(id, orElseUpdateAndGet(id))

    /**
     * find by id
     */
    fun findOrNull(id: ID): T? = store[id]?.data

    /**
     * update
     * this functions ignore cache time
     */
    fun update(id: ID, data: T): T = data.apply { store[id] = CacheWrapper(System.currentTimeMillis() + cacheTimeMillis, this) }

    /**
     * remove
     */
    fun remove(id: ID): T? = store.remove(id)?.data

    /**
     * clear all data
     */
    fun clear(): Unit = store.clear()

    /**
     * foreach key-data
     */
    fun forEach(action: BiConsumer<ID, T>) = store.forEach { (k, v) -> v.data?.also { action.accept(k, it) } }

    /**
     * all data
     */
    fun all(): Map<ID, T> = HashMap<ID, T>().also { forEach { k, v -> it[k] = v } }

    /**
     * all keys
     */
    fun keys() = ArrayList<ID>(store.size).also { forEach { k, _ -> it.add(k) } }

    private class CacheWrapper<T> (
        private val expireTimeMillis: Long,
        private val _data: T
    ) {
        val data get(): T? = if (expireTimeMillis >= System.currentTimeMillis()) _data else null
    }
}

